sdks/java/core/src/test/java/org/apache/beam/sdk/io/TextIOReadTest.java

```diff
@@ -214,6 +214,38 @@ private void runTestRead(String[] expected) throws Exception {
 
   @Test
   @Category(NeedsRunner.class)
+  public void testReadStringsWithCustomSeparator() throws Exception {
+    final String[] inputStrings = new String[] {
+        // incomplete separator
+        "To be, or not to be: that |is the question: ",
+        // complete separator
+        "Whether 'tis nobler in the mind to suffer ||",
+        //truncated separator
+        "The slings and arrows of outrageous fortune,|" };
+
+    File tmpFile = Files.createTempFile(tempFolder, "file", "txt").toFile();
+    String filename = tmpFile.getPath();
+
+    try (PrintStream writer = new PrintStream(new FileOutputStream(tmpFile))) {
+      for (String elem : inputStrings) {
+        byte[] encodedElem = CoderUtils.encodeToByteArray(StringUtf8Coder.of(), elem);
+        String line = new String(encodedElem);
+        writer.print(line);
+      }
+    }
+    String[] expected = new String[] {
+        "To be, or not to be: that |is the question: Whether 'tis nobler in the mind to suffer ",
+        "The slings and arrows of outrageous fortune,|" };
+    TextIO.Read read = TextIO.read().from(filename).withSeparator(new byte[] {'|', '|'});
+
+    PCollection<String> output = p.apply(read);
+
+    PAssert.that(output).containsInAnyOrder(expected);
+    p.run();
+  }
+
```

Comment:
Please add a few tests using SourceTestUtils.assertSplitAtFractionExhaustive, similar to testSplittingSourceWithMixedDelimitersAndNonEmptyBytesAtEnd and ones around it.

I tried adding one such test and it failed - implementing this properly is very tricky, especially if you allow arbitrary delimiters that can self-overlap, such as `||` in your example: how should we interpret `abc|||xyz` - as `abc|`, `xyz` - or as `abc`, `|xyz`? And how do we consistently enforce this interpretation if the file is split by the runner into chunks differently?

Reply:
Good points, thanks for pointing me at them! I'll try to add those tests. It is tricky indeed. I understand that you did not have the overlap problem with` \r` and `\n` because any combination of `\r` and `\n` , no matter where we split, will result in a split  and no `\r` or `\n` in the records. This is not the case with fixed custom multi-byte separator as you pointed. I'll think about this problem.

Reply:
When I manage to have a coherent delimiter management no matter the faction splitting point, `||||` should produce and empty record right?

Reply:
I have fixed the parser to work no matter what the source split point is and I have added the corresponding test with the custom separator. 
What I have now to support is the overlap that you were talking about. Obviously I cannot rewind the offset of `(separator.size + 1)` to allow only one byte overlap, neither can I rewind the offset of `(2*sperator.size)` to allow maximum overlap because it might produce duplicate record if a `record.size < separator.size`. I cannot either catch anything to state that the file format is wrong in case of overlap because I will get no exception, just flaky record content depending on the runner / source split point. Do you have any idea on how to support overlap of separator?

Reply:
How about simply prohibiting separators that can have self-overlap? A separator can have self-overlap, and hence cause ambiguous parsing, if it has a suffix that is also its prefix - i.e. there exists "i" < separator.length such that the last "i" bytes == the first "i" bytes - can simply check that in the "withSeparator" method using a quadratic algorithm cause separator is unlikely to be more than a couple of bytes.

Reply:
yes, I like this, it's simple and efficient. still, we have (odd) client use cases where separator might be `\n xxxxxx \n` but in worse cases 10 bytes, so adding a verification algorithm that is O(10²) and that is run only once at pipeline construction time will do. Thanks.